lib/client.js

```diff
@@ -112,34 +152,60 @@ module.exports = function (dependencies) {
 
-    return new Promise ( resolve => {
+    return new Promise ( resolve => {      
       request.on("end", () => {
+        try {
+          if (logger.enabled) {
+            logger(`Request ended with status ${status} and responseData: ${responseData}`);
+          }
+  
+          if (status === 200) {
+            resolve({ device });
+          } else if (responseData !== "") {
+            const response = JSON.parse(responseData);
+  
+            if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
+              this.config.token.regenerate(tokenGeneration);
+              resolve(this.write(notification, device, retryCount + 1));
+              return;
+            } else if (status === 500 && response.reason === "InternalServerError") {
+              this.closeAndDestroySession();
+              let error = new VError("Error 500, stream ended unexpectedly");
+              resolve({ device, error });
+              return;
+            }
+  
+            resolve({ device, status, response });
+          } else {
+            let error = new VError("stream ended unexpectedly");
+            resolve({ device, error });
+          }
+        } catch (e) {
+          const error = new VError(e, 'Unexpected error processing APNs response');
+          if (logger.enabled) {
+            logger(`Unexpected error processing APNs response: ${e.message}`);
+          }
+          resolve({ device, error });
+        }
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
         if (logger.enabled) {
-          logger(`Request ended with status ${status} and responseData: ${responseData}`);
+          logger('Request timeout');
         }
 
-        if (status === 200) {
-          resolve({ device });
-        } else if (responseData !== "") {
-          const response = JSON.parse(responseData);
+        request.close(NGHTTP2_CANCEL);
 
-          if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
-            this.config.token.regenerate(tokenGeneration);
-            resolve(this.write(notification, device, retryCount + 1));
-            return;
-          } else if (status === 500 && response.reason === "InternalServerError") {
-            this.session.destroy();
-            let error = new VError("Error 500, stream ended unexpectedly");
-            resolve({ device, error });
-            return;
-          }
+        resolve({ device, error: new VError("apn write timeout") });
+      });
 
-          resolve({ device, status, response });
-        } else {
-          let error = new VError("stream ended unexpectedly");
-          resolve({ device, error });
+      request.on("aborted", () => {
+        if (logger.enabled) {
+          logger('Request aborted');
         }
-      })
+
+        resolve({ device, error: new VError("apn write aborted") });
+      });
 
       request.on("error", (error) => {
         if (logger.enabled) {
```

Comment:
I noticed an error when running typescript to see if there were similar issues to those I had already seen.

**This may cause the promise to hang if an apn write fails - this will probably throw a ReferenceError due to the undefined variable**

```patch
         if (typeof error === "string") {
-          error = new VError("apn write failed: %s", err);
++          error = new VError("apn write failed: %s", error);
```

I noticed this using typescript (some options in tsconfig.json are unnecessary/inapplicable)

```js
// tsconfig.json
{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "outDir": ".built",
    "allowJs": true,
    "checkJs": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  },
  "exclude": ["index.d.ts", ".built"]
}
```

Reply:
https://nodejs.org/api/http2.html#http2_class_http2stream claims that the 'error' should be an instance of `Error`, though, so hopefully not the cause of any issues.

LGTM otherwise

Reply:
https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js seems like it doesn't enforce that the error is an object, e.g. if the caller of destroy() passes in a string instead of an Error object, but it doesn't seem as if this is the reason for node-apn crashing in practice

Reply:
https://github.com/TysonAndre/node-apn/commit/60b760d896d7712c64d88146f062dd4e02df83f0 fixes that and some nits on trailing whitespace

Reply:
done